early draft of foundational FVM FIP

raulk opened the pull request.

raulk committed:
skeleton of Foundational FVM FIP.

raulk committed:
flesh out the FIP.

raulk committed:
update ToC.

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+User-programmability is the ability to deploy and run arbitrary user-provided
+code on some infrastructure, in this case a blockchain network.
+
+Today, the Filecoin blockchain lacks user-programmability. This FIP is part of a
```
Well, on-chain user programmability. Users _can_ deploy custom logic, just not on-chain.

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+
+A change of this magnitude impacts many components. Concretely:
+
+1. The storage layer, to store and load user-provided code.
```
I'd be clear what you mean by "storage" here. Maybe "chain state"?

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+
+We specify the Filecoin Virtual Machine (FVM), a IPLD-ready WASM-based execution
+layer capable of running arbitrary user-provided code. The FVM replaces the
+existing execution layer.
```
I'd define the existing layer. Maybe just link to specs-actors?

anorth replied:
Or rather the existing spec's description of the VM?

raulk replied:
Added a link to https://spec.filecoin.io/intro/filecoin_vm/. I think it's strictly more correct than linking to specs-actors, as the new VM does not _replace_ specs-actors.

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+
+We provide a reference architecture for the FVM, and explain its various
+components: (a) Machine, (b) Call Manager, (c) Invocation Container, (d) Kernel,
+(e) Syscalls, and (f) Externs. We define the contract that actor code must
```
I'm not sure if `Externs` have a place here. We've implemented the reference FVM that way, but a large part of that is the rust/go boundary.

raulk replied:
I think of the Externs as a logical boundary; it's a set of externally provided functions that the VM needs to operate. Indeed it is emphasised in the Go/Rust integration because in that case it becomes a _physical_ boundary due to the cross of technologies. I'll clarify this aspect in the doc, and we can continue to debate on the public FIP if that SGTY?

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+adhere to, the catalogue of available syscalls, the mechanics of state
+manipulation, the error conditions, and more technical details.
+
+In this FIP, we propose a network upgrade involving two major changes: (1) the
+introduction of a non-programmable version of the FVM to pave the way, and (2)
```
For 1, I'd specify that it's non-consensus-breaking.

raulk replied:
Ok, I can add that qualifier. One question, then: we don't plan to introduce exit code breaking changes at this stage (M1)? Only at M2? I was under the impression that you really wanted to introduce some minor rationalization in M1, but we're just not doing that for M0 (because we are mainnet canaries at this stage).

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+
+## Change Motivation
+
+The desire for user-programmability in Filecoin is well-recognized and
+uncontested. It is the key to unlock enormous latent potential, as it enables
```
I wouldn't say "uncontested", that's inviting someone to challenge you.

raulk replied:
Ok, I'll lower the tone :-)

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
````diff
@@ -0,0 +1,1010 @@
+A VM is instantiated with these input parameters:
+
+```rust
+let machine = Machine::new(
````
This could confuse readers a bit given that the ref-fvm currently distinguishes between the "executor" (the thing that executes messages) and the "machine" (the thing on which messages are executed).

raulk replied:
I'll add a note.

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+- 4GiB addressable memory not deemed limiting for blockchain use cases
+- No loss of function: 64-bit integer types and operations are supported
+
+The FVM requires [multi-value support]. It explicitly forbids [SIMD
```
I'd make it clear that these are the _current_ requirements and that they might change.

Stebalien replied:
For example, SIMD & floats are useful as long as they're deterministic. SIMD because it means we may be able to have fewer syscalls, and floats because statically forbidding floats can be a bit tricky.

anorth replied:
Yeah SIMD as a future option would be great.

raulk replied:
Clarified in that regard.

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+
+The FVM is responsible for these tasks:
+
+- Loading the code of invoked actors, turning them into WASM modules, and
+  optimizing the process through compilation and caching.
```
> turning them into WASM modules

This is confusing, the code is _already_ a wasm module before it gets to the FVM. I believe you mean "loading on-chain wasm modules and compiling them down to native code"?

raulk replied:
Yep, I was thinking too close to the code, i.e. the instantiation of a `wasmtime::Module` from bytecode.

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+
+### Responsibilities
+
+The FVM is responsible for these tasks:
```
I think we can simplify this to something like:

> The FVM applies messages to the state-tree. It is not responsible for advancing the chain itself.

I feel that enumerating the responsibilities may just confuse more than it helps.

raulk replied:
I feel like this enumeration is useful as an overview of everything that the FVM does -- it might be obvious to us but might not be to folks that are less familiar with the details. I'll keep it for now, but if it squeaks again in the public review, I'm happy removing.

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+
+### Actors
+
+The term _Actor_ is a reference to the [actor model], a concurrent computation
+paradigm that inspires Filecoin's programming and scalability philosophies.
```
I... wouldn't mention the actor model here given that we've basically ignored it entirely. I'd just say that an actor is equivalent to a smart contract.

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+Filecoin client. The recommended approach is to embed a CAR bundle with all the
+relevant bytecode, and seeding the blockstore with it on node start.
+
+**Canonical codebase**
+
+With the definition of a universal, portable execution bytecode format, all
+Filecoin clients can rely on a single codebase for built-in actors. This speeds
+up protocol development, shortens protocol upgrade timelines, and reduces
+coordination overhead.
+
+This FIP proposes to promote the FVM built-in actors codebase (forked from the
+Forest client), to the status of **_canonical_ built-in actors**. Clients should
+adopt this codebase, but they may choose to pursue their own.
```
This isn't quite correct. Everyone _must_ adopt the "canonical" actors, otherwise gas accounting will be wrong.

raulk replied:
True.

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+Forest client), to the status of **_canonical_ built-in actors**. Clients should
+adopt this codebase, but they may choose to pursue their own.
+
+##### Native actors
```
This is going to be moderately confusing as people may assume "native" means "native code" (i.e., runs directly on the host).

raulk replied:
True, we need a better name for this.

raulk replied:
I can't think of a better name now. Let's see how much it confuses folks when it's publicly submitted.

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+stdlibs, etc.) may lead to oversized WASM bytecode and execution overheads,
+translating to higher gas costs.
+
+Furthermore, there may be hard limits enforced on the sizing of bytecode
+deployables in the future.
```
"will likely be" and "before user deployable actors are allowed".

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+There's a [reference SDK] written in
+Rust to accompany this FIP. It is used by the canonical built-in actors.
+
+We find that Rust is a reasonable choice to writing native actors due to the
```
Well... we're currently using std. Although I guess we'd like to remove it.

raulk replied:
Indeed.

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+taking two input parameters:
+
+1. The actor's WASM bytecode as an input parameter.
+2. The invocation container version.
```
IMO, the wasm module should just specify this. We can add custom fields to the wasm metadata. Or, we could just version the runtime methods themselves.

raulk replied:
Do you mean as an exported global? That would work as long as there are no cases where we need to know the version _before_ compiling the module. (We _could_ locate the designated export by scanning the bytecode if we need to do that, but that's probably wasteful; we'd want to compile and then enumerate the exports in the wasmtime::Module).

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+    - Syntax validation and structural validation, [as per standard](https://webassembly.github.io/spec/core/valid/index.html).
+    - No floating point instructions.
+2. Multihash the bytecode and calculate the code CID.
+3. Check if the code CID is already registered in the `ActorRegistry`.
+4. If yes, charge gas for the above operations and return the existing CID.
+5. Insert an entry in the `ActorRegistry`, binding the bytecode CID with the
```
Could maybe simplify this by appealing to idempotency, but I'm not sure if that'll help.

raulk replied:
IMO mentioning idempotency here may lead to confusion because the _full_ logic is not idempotent due to the lower gas on duplicate.

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+6. Return the CID, and charge the cost of the above operations, and potentially
+   a price for rent/storage.
+
+At this point, the actor will be ready to be instantiated through the standard
+`InitActor#Exec`. Any params provided will be passed through to the actor's
+constructor (method number = 0). This includes source depoyable byte streams
+(EVM bytecode) for foreign actors.
```
Importantly, actors can be deployed once and instantiated many times.

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+- 
+- The `Result` return type 
+
+#### Namespace: `actor`
```
I'm not sure if we should bother specifying these for now as they're designed for a "trusted" system. We'll probably be better off if we keep _this_ FIP to higher-level abstractions, introducing specific design decisions in future FIPs.

raulk replied:
My intention was to provide sufficient level of detail for (alternative) FVM implementations and/or SDK developers to implement off the spec. This is generally the level of detail that is submitted in equivalent documents in other communities (e.g. EIP).

raulk replied:
Oh, you're referring to the `actor` namespace specifically. IMO the only syscall that doesn't belong long-term is `create_actor`. Everything else will probably remain, won't it?

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+
+### IPLD memory model
+
+IPLD is the data model of Filecoin. Actors can be construed as logic that
+receives an input IPLD graph, performs computation, and (optionally) mutates the
+IPLD graph by saving an alternative root in the actor's state.
```
nit: "and returns a new IPLD graph".

I'd avoid the word "mutate" as IPLD is copy-on-write.

raulk replied:
Works for me, I had worded it that way before, but it's not technically accurate today. However, it is conceptually correct.

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+node side, and it is exposed to the FVM via an Extern. This is important as the
+VM itself needs to understand the IPLD so we can do garbage collection, etc.
+
+IPLD codecs will be defined as WASM libraries, imported through the WASM import
+system. Ideally, ADLs and selectors would be implemented in the same way.
```
I'd drop this statement for now (raises more questions than it answers).

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+tree. This is done through the maintenance of an "accessible set" inside the
+kernel.
+
+Because Filecoin IPLD state objects are highly atomized and linked, accessing
+and mutating entries in objects like HAMTs and AMT results in multiple
+sequential state IO operations, each of which traverses the Extern boundary in a
+non-parallelizable way. If the Extern is traversed through FFI, the cost of
+operating on ADLs may be non-negligible.
+
+Alternative strategies like moving ownership of the state store to the FVM-side,
+or leveraging selectors for queries, may be explored in the future.
```
This seems implementation specific. I'd leave this off as well.

anorth commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+  of richer, composite, interconnected solutions.
+- Unification of built-in actors across implementations to speed up protocol
+  development.
+- Enabling automatic/governance-driven protocol upgrades.
```
I don't quite understand this one. Is it referring to user-programmed contracts having their own governance, detached from core protocol, and upgrading thusly? Or is it something about implementing Filecoin core protocol upgrades via FVM?

raulk replied:
Yeah, this is way too concise to actually mean anything concrete. I'll expand. It was more oriented towards the second element (the ability to deploy Filecoin core protocol upgrades that affect only actor logic as governance-driven fork activations).

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+Alternative strategies like moving ownership of the state store to the FVM-side,
+or leveraging selectors for queries, may be explored in the future.
+
+### Error conditions
```
I'd drop this. We can mention that syscalls have "error numbers" in the syscall section, and message exit codes are already defined.

anorth commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+
+Once the caller has finished applying all messages, it flushes the machine. This
+computes the final state root. Depending on implementation, this operation may
+also commit newly written blocks to the blockstore.
```
May be worth noting more explicitly that the machine is stateful, what state it holds, and why. Is it in practise a caching-style optimisation, but equivalent to instantiating a new machine for each message and threading the state roots between them?

raulk replied:
Yes, this is mostly a batching mechanism for mechanical sympathy, as the most common use case is to apply a batch of ordered messages as they appear in a tipset. I'll expand.

Stebalien commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+
+### Memory policy
+
+The Invocation Container makes available a maximum of 1024 pages to the actor.
+Given that WASM memory pages are a fixed 64KiB size, this results in every actor
+call being limited to 64MiB of memory.
+
+Depending on the WASM engine, it may be possible to size memory dynamically. If
+that's the case, a policy that grows memory dynamically from a low value (even
+zero) should be preferred.
```
I'd remove the details here as they're far from settled (and there are a lot of nuances involving wasm stack frames/depth, globals, locals, etc.).

The memory limit will also have to apply to the entire message execution. Otherwise the maximum would be max_depth * limit which is 64*4096 (I think?).

raulk replied:
I'll remove. We need to keep a holding pen document with unspecified details to figure out, though.

Stebalien reviewed (commented):
Overall, this seems ready for a public PR even if it's not completely done.

The main thing I'd do is remove implementation details that aren't settled. I'm concerned about lots of debate over bits that aren't done and/or aren't relevant to the network, distracting from the main goal of the proposal.

anorth commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+- Widespread support
+- Lower memory footprint
+- 4GiB addressable memory not deemed limiting for blockchain use cases
+- No loss of function: 64-bit integer types and operations are supported
```
Interesting. This might be a good discussion to have in public to flush out more possible use cases.

This might limit possible universality of FVM as an execution environment outside the blockchain state transition function. For example, an FVM rollup could have far greater demands on hardware, including memory, to enable much more complex and resource-intensive programs to be written, while retaining Filecoin's blockchain security. Similarly, for compute-over-data projects there's a potential world in which something very like the FVM is the target for code to run inside compute providers, and support very large memory execution environments. It could be very powerful to have seamless, binary interop between rollups and off-chain compute. Taken to the extreme, it might be that we develop a zk-FVM which could do both.

I get the current requirements are rightly focussed on the blockchain state use case, but is there a big gain there from taking the 32-bit route?

raulk replied:
Agree, this is a good discussion to have openly. Wasm was conceived with 32-bit pointers, and 64-bit addresing is being introduced through the [memory64 extension](https://github.com/WebAssembly/memory64/blob/main/proposals/memory64/Overview.md), which has just graduated to [Phase 3](https://twitter.com/wvo/status/1362458368107941888), but is not yet part of the core spec. We're out of options here, other than following closely and potentially forking the FVM into a bleeding-edge version that generalises the programming environment for use cases beyond blockchain state function execution. Things like SIMD and memory64 could be experimented with there.

anorth commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+
+Because this contract may change over time, user-deployed actors may specify an
+IC version at [`InitActor#InstallActor` time](#actor-deployment). The strategy
+for deprecating IC versions over time has not been yet determined.
```
Nice

anorth commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+- The single uint32 argument corresponds to the block ID of the encoded input
+  parameters struct, or `0` for none.
+- The single uint32 return value corresponds to the block ID of the return data
+  struct, or `0` for none.
```
I don't understand what a block ID is. Could you describe it or point to a reference in the text?

raulk replied:
Added more details here: https://github.com/raulk/FIPs/pull/1/commits/53323f0239d9ab07440eb215a75d5d29b4f148ae.

anorth commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+
+The _method number_ field on messages remains unchanged to preserve backwards
+compatibility. It may be deprecated and eventually removed later. Built-in
+actors continue relying on _method number_, but now perform internal dispatch.
```
Could you give an example of what the params block (?) looks like for a built-in actor method that takes a struct parameter? Is the block a struct that has method number and parameter embedded? 

And to be concrete, can other actors use a string in their params block for method dispatch?

I think internal dispatch is going to be great, just struggling to see the implementation from this brief description, and it's important to things like contract API standards, etc.

anorth commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+) -> Result<u32>;
+
+/// Verifies a sector seal proof.
+pub fn verify_seal(info_off: *const u8, info_len: u32) -> Result<i32>;
```
What's the output here? Does it throw when it fails? Ditto for some others below.

anorth commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+pub fn receiver() -> Result<u64>;
+
+/// Returns the method number from the message.
+pub fn method_number() -> Result<u64>;
```
What's this needed for? Is this how the built-in actors do internal dispatch? Do we expect zero by convention for most other actors?

raulk replied:
Indeed, this is here for backwards compatibility. I'm not sure if `0` is a safe default to return to imply "no method number" (if we made this value optional in the Message going forward), as this currently maps to the "send method", although that's whole situation needs to be revisited too.

anorth commented on FIPS/fip-draft_foundational_fvm.md:
````diff
@@ -0,0 +1,1010 @@
+#### Namespace: `self`
+
+```rust
+/// Gets the current root for the calling actor.
````
State root? I.e. `ActorState.state`, not the actor root that also contains the code CID, nonce etc.

raulk replied:
Yep, I don't think we call/refer to the CID of the `ActorState` object (which is what I think you're referring to) anything special. But I'll clarify.

raulk replied:
Added the clarification.

anorth commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+
+The Invocation Container makes available a maximum of 1024 pages to the actor.
+Given that WASM memory pages are a fixed 64KiB size, this results in every actor
+call being limited to 64MiB of memory.
```
Is there likely to be any cost to using the memory, or does each actor have that max allocation for free? (I know details will come in a follow-up).

raulk replied:
This is a great question and I think we should move it to the public discussion.

anorth commented on FIPS/fip-draft_foundational_fvm.md:
```diff
@@ -0,0 +1,1010 @@
+  creating an account, as per above.
+- Power actor and Storage Market actor: state is queried when handling the
+  `network::total_fil_circ_supply` syscall.
+- Burnt funds account and Reserve account: balances are queried when handling
```
Please lmk if there are any change we could make here that would make things significantly easier for you.

anorth reviewed (commented):
This is great, thanks for expressing the detail here so clearly.

In `actor-execution.png`, the lower-right `WASM` label confuses me. It seems to suggest that the responsibilities of that yellow block are implemented in WASM. I expected them to be implemented natively, but _provide_ a WASM execution environment. Am I wrong?

raulk commented:
@anorth

> I expected them to be implemented natively, but provide a WASM execution environment. Am I wrong?

Well noticed. I had corrected this for the FVM website (even though the labels were ultimately cropped off due to design constraints), but forgot to correct it here.
![image](https://user-images.githubusercontent.com/1101242/152073494-cf2a47e7-27ce-41cb-a580-a5d9109de076.png)

raulk committed:
apply review comments.

raulk committed:
address more review comments.

raulk commented:
@Stebalien I've addressed all your comments, and resolved the threads where I accepted the suggestion. I replied and kept the thread open where I'd appreciate a second round of input.

raulk committed:
clarify wasm32 vs wasm64; rm memory policy.

raulk committed:
add more details on block ID and ipld syscall mechanics.

raulk committed:
clarify self::root syscall.

raulk committed:
improve syscall specs.

raulk committed:
update actor-execution.png.

raulk committed:
more polish.
